compiler/rustc_hir_analysis/src/variance/constraints.rs

```diff
@@ -252,10 +261,18 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
                 self.add_constraints_from_args(current, def.did(), args, variance);
             }
 
-            ty::Alias(_, ref data) => {
+            ty::Alias(ty::Projection | ty::Inherent | ty::Opaque, ref data) => {
                 self.add_constraints_from_invariant_args(current, data.args, variance);
             }
 
+            ty::Alias(ty::Weak, ref data) => {
+                if self.tcx().features().lazy_type_alias {
+                    self.add_constraints_from_args(current, data.def_id, data.args, variance);
+                } else {
+                    self.add_constraints_from_invariant_args(current, data.args, variance);
+                }
+            }
```

Comment:
I'm not sure if this is correct if the `ty` comes from an external crate. Ideally (?) we'd use `add_constraints_from_args` for *any* weak projection.